Fix: Preserve map state when switching between views

ngoiyaeric opened the pull request:
### **User description**
This commit addresses an issue where the map would lose its state (drawings, location, zoom) when you navigated between the map view and the settings view.

The fix involves the following changes:

1.  **Enhanced `MapDataContext`**:
    *   The `MapDataContext` now stores the Mapbox GL JS map instance itself, the current viewport details (center, zoom, pitch, bearing), in addition to the already existing `drawnFeatures`.

2.  **Modified `Mapbox` Component Lifecycle**:
    *   The `Mapbox` component now sources its map instance and viewport state from `MapDataContext`.
    *   On mount, if a map instance exists in the context, it's reused. Its canvas is re-appended to the `Mapbox` component's container div, the map is resized, and its viewport is restored. Event listeners and drawing tools are re-initialized.
    *   If no instance exists in context, a new map is created and stored in the context.
    *   The `useEffect` cleanup in `Mapbox` no longer calls `map.remove()`, allowing the map instance to persist in the context. It now only detaches component-specific event listeners.
    *   The `captureMapCenter` function now updates the viewport details in `MapDataContext`.

3.  **Drawing Restoration**:
    *   The `setupDrawingTools` function was updated to prioritize loading features from `mapData.drawnFeatures` (stored in context) when the drawing tools are initialized, ensuring that drawings persist across component unmounts/remounts.

4.  **Build Verification**:
    *   The changes were verified to allow the project to build successfully using `bun run build`.

These changes ensure that your map interactions, including drawings and map position, are maintained seamlessly when toggling between the map and other views like settings.


___

### **PR Type**
Bug fix, Enhancement


___

### **Description**
• Fix map state persistence when switching between views
• Preserve map instance, viewport, and drawings across component unmounts
• Enhanced MapDataContext to store map instance and viewport state
• Modified Mapbox component lifecycle to reuse existing map instances


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>map-data-context.tsx</strong><dd><code>Enhanced context to store map instance and viewport</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

components/map/map-data-context.tsx

• Added <code>mapInstance</code> field to store Mapbox GL JS map instance<br> • Added <br><code>viewport</code> field to store center, zoom, pitch, and bearing<br> • Updated <br>initial state with default viewport values<br> • Imported <code>mapboxgl</code> for <br>type definitions


</details>


  </td>
  <td><a href="https://github.com/QueueLab/QCX/pull/190/files#diff-962700610dd525e2640fcbf6400b4f1486142becccbca3b6e0c941e42ef5af6a">+13/-2</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>mapbox-map.tsx</strong><dd><code>Refactored map lifecycle for state persistence</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

components/map/mapbox-map.tsx

• Modified map initialization to reuse existing instance from context<br> <br>• Updated cleanup to preserve map instance instead of removing it<br> • <br>Enhanced drawing restoration to prioritize context-stored features<br> • <br>Updated <code>captureMapCenter</code> to store viewport in context<br> • Restructured <br>component lifecycle for map state persistence


</details>


  </td>
  <td><a href="https://github.com/QueueLab/QCX/pull/190/files#diff-ce7ea3ef85a6812c7da39941ffa47e8e0fadbb3fc7c391cf1cafd96303cf3a0f">+193/-107</a></td>

</tr>
</table></td></tr><tr><td><strong>Additional files</strong></td><td><table>
<tr>
  <td><strong>install_bun.sh</strong></td>
  <td><a href="https://github.com/QueueLab/QCX/pull/190/files#diff-8fbeb98b6e648aeb9d131c3a74d0fc2fe3cda4256f762a4fb8833e3679b20125">[link]</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about Qodo Merge usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- The map now maintains and restores its viewport settings (center, zoom, pitch, bearing) for a more consistent user experience.
- **Refactor**
	- Improved map initialization and cleanup processes for better reliability and performance.
- **Chores**
	- Minor code and formatting adjustments for maintainability.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

google-labs-jules[bot] committed:
Fix: Preserve map state when switching between views

This commit addresses an issue where the map would lose its state (drawings, location, zoom) when you navigated between the map view and the settings view.

The fix involves the following changes:

1.  **Enhanced `MapDataContext`**:
    *   The `MapDataContext` now stores the Mapbox GL JS map instance itself, the current viewport details (center, zoom, pitch, bearing), in addition to the already existing `drawnFeatures`.

2.  **Modified `Mapbox` Component Lifecycle**:
    *   The `Mapbox` component now sources its map instance and viewport state from `MapDataContext`.
    *   On mount, if a map instance exists in the context, it's reused. Its canvas is re-appended to the `Mapbox` component's container div, the map is resized, and its viewport is restored. Event listeners and drawing tools are re-initialized.
    *   If no instance exists in context, a new map is created and stored in the context.
    *   The `useEffect` cleanup in `Mapbox` no longer calls `map.remove()`, allowing the map instance to persist in the context. It now only detaches component-specific event listeners.
    *   The `captureMapCenter` function now updates the viewport details in `MapDataContext`.

3.  **Drawing Restoration**:
    *   The `setupDrawingTools` function was updated to prioritize loading features from `mapData.drawnFeatures` (stored in context) when the drawing tools are initialized, ensuring that drawings persist across component unmounts/remounts.

4.  **Build Verification**:
    *   The changes were verified to allow the project to build successfully using `bun run build`.

These changes ensure that your map interactions, including drawings and map position, are maintained seamlessly when toggling between the map and other views like settings.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/QueueLab/QCX?pullRequest=190) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/QueueLab/QCX?pullRequest=190) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/QueueLab/QCX?pullRequest=190) it.</sub>

google-labs-jules[bot] committed:
Fix: Ensure map interactivity and rendering on remount

This commit further addresses issues where the map would become a non-interactive image or lose context when returning from other views (e.g., settings). This builds upon the previous fix that preserved the map instance.

The key changes include:

1.  **Robust Map Re-initialization**:
    *   When reusing an existing map instance after its component remounts:
        *   The map's canvas is re-appended to its container.
        *   A deferred `resize()` call (100ms `setTimeout`) allows the DOM to settle.
        *   The viewport (center, zoom, pitch, bearing) is restored *after* the initial resize within the timeout.

2.  **Conditional Style Reload**:
    *   If, after remounting and initial resize, `map.isStyleLoaded()` returns `false`, the map's style is reloaded using `map.setStyle(currentStyle)`.
    *   After the style is confirmed to be processed (via the `styledata` event):
        *   Essential map elements like DEM terrain and the sky layer are re-added.
        *   The viewport is restored again.
        *   Drawing tools are re-initialized (which includes restoring drawn features from context).
        *   The map is resized again to ensure correctness after all additions.

3.  **PreserveDrawingBuffer Investigation**:
    *   Confirmed that `preserveDrawingBuffer` is not explicitly set and defaults to `false`, which is appropriate and not contributing to the issue.

4.  **Build Verification**:
    *   All changes were verified to build successfully.

These enhancements aim to ensure that the map not only preserves its data but also correctly re-renders and remains fully interactive when its component is hidden and then shown again.

google-labs-jules[bot] committed:
Refactor: Always create fresh map instance to ensure correct rendering

This commit addresses the issue where the map would render as a static image or lose interactivity when its component was re-mounted (e.g., when switching between settings and map views).

The core change is a new strategy for managing the Mapbox GL JS map lifecycle:

1.  **No Map Instance in Context**:
    *   `MapDataContext` no longer stores the `mapboxgl.Map` instance. It now only stores serializable map data: viewport (center, zoom, pitch, bearing) and drawn features.

2.  **Fresh Map on Mount**:
    *   The `Mapbox` component now *always* creates a new `mapboxgl.Map` instance when it mounts.
    *   The complex logic for trying to reuse and revive a previously stored map instance has been removed.

3.  **Initialization from Context Data**:
    *   The newly created map is initialized using the `viewport` data from `MapDataContext`.
    *   Drawn features are restored onto the new map from `mapData.drawnFeatures` after the map loads and drawing tools are set up.

4.  **Consistent Cleanup**:
    *   Each `Mapbox` component instance now fully owns its map. The `useEffect` cleanup function always calls `map.current.remove()` to destroy the instance when the component unmounts.

5.  **Build Verification**:
    *   All changes were verified to build successfully using `bun run build`.

This approach avoids potential issues with stale WebGL contexts or corrupted internal state in a long-lived map instance whose canvas is repeatedly detached and re-attached. By creating a fresh map instance each time, we rely on Mapbox GL JS's standard initialization process, which should lead to more reliable rendering and interactivity.

google-labs-jules[bot] committed:
Refactor: Always create fresh map instance to ensure correct rendering

This commit addresses the issue where the map would render as a static image or lose interactivity when its component was re-mounted (e.g., when switching between settings and map views).

The core change is a new strategy for managing the Mapbox GL JS map lifecycle:

1.  **No Map Instance in Context**:
    *   `MapDataContext` no longer stores the `mapboxgl.Map` instance. It now only stores serializable map data: viewport (center, zoom, pitch, bearing) and drawn features.

2.  **Fresh Map on Mount**:
    *   The `Mapbox` component now *always* creates a new `mapboxgl.Map` instance when it mounts.
    *   The complex logic for trying to reuse and revive a previously stored map instance has been removed.

3.  **Initialization from Context Data**:
    *   The newly created map is initialized using the `viewport` data from `MapDataContext`.
    *   Drawn features are restored onto the new map from `mapData.drawnFeatures` after the map loads and drawing tools are set up.

4.  **Consistent Cleanup**:
    *   Each `Mapbox` component instance now fully owns its map. The `useEffect` cleanup function always calls `map.current.remove()` to destroy the instance when the component unmounts.

5.  **Build Verification**:
    *   All changes were verified to build successfully using `bun run build`.

This approach avoids potential issues with stale WebGL contexts or corrupted internal state in a long-lived map instance whose canvas is repeatedly detached and re-attached. By creating a fresh map instance each time, we rely on Mapbox GL JS's standard initialization process, which should lead to more reliable rendering and interactivity.

google-labs-jules[bot] committed:
Fix: Prevent Lottie animation rapid re-render loop

This commit addresses an issue where a Lottie animation (used for map loading state) would re-render rapidly, making the system unusable.

The root cause was identified as a loop in the `Mapbox.tsx` component:
1. The main `useEffect` responsible for map initialization depended on `mapData.viewport`.
2. If `mapData.viewport` was initially undefined, this effect would calculate a default viewport and save it to `MapDataContext` using `setMapData`.
3. This update to `mapData.viewport` would then cause the same `useEffect` to re-run (due to the dependency), leading to its cleanup phase being called (which set `isMapLoaded` to `false`) and then its setup phase re-executing (which eventually set `isMapLoaded` to `true`).
4. This rapid toggling of `isMapLoaded` caused the `ConditionalLottie` component to flash the animation.

The fix involves:
- Removing `mapData.viewport` from the dependency array of the main map initialization `useEffect` in `Mapbox.tsx`.
- The effect still determines and sets the initial default viewport in `MapDataContext` if it's undefined, but this action no longer causes the effect itself to be torn down and re-run. A guard condition `if (map.current || !mapContainer.current)` within the effect prevents re-initialization of the map during the re-render caused by `setMapData`.

An ESLint warning for `exhaustive-deps` regarding the omitted `viewport` dependency is acknowledged; this omission is intentional and necessary for the fix.

The project builds successfully with these changes. This should resolve the Lottie animation loop and improve system usability.

google-labs-jules[bot] committed:
Fix: Prevent Lottie animation rapid re-render loop

This commit addresses an issue where a Lottie animation (used for map loading state) would re-render rapidly, making the system unusable.

The root cause was identified as a loop in the `Mapbox.tsx` component:
1. The main `useEffect` responsible for map initialization depended on `mapData.viewport`.
2. If `mapData.viewport` was initially undefined, this effect would calculate a default viewport and save it to `MapDataContext` using `setMapData`.
3. This update to `mapData.viewport` would then cause the same `useEffect` to re-run (due to the dependency), leading to its cleanup phase being called (which set `isMapLoaded` to `false`) and then its setup phase re-executing (which eventually set `isMapLoaded` to `true`).
4. This rapid toggling of `isMapLoaded` caused the `ConditionalLottie` component to flash the animation.

The fix involves:
- Removing `mapData.viewport` from the dependency array of the main map initialization `useEffect` in `Mapbox.tsx`.
- The effect still determines and sets the initial default viewport in `MapDataContext` if it's undefined, but this action no longer causes the effect itself to be torn down and re-run. A guard condition `if (map.current || !mapContainer.current)` within the effect prevents re-initialization of the map during the re-render caused by `setMapData`.

An ESLint warning for `exhaustive-deps` regarding the omitted `viewport` dependency is acknowledged; this omission is intentional and necessary for the fix.

The project builds successfully with these changes. This should resolve the Lottie animation loop and improve system usability.

google-labs-jules[bot] committed:
Fix: Restore drawing calculations by preventing map re-creation loop

This commit addresses an issue where your drawing calculations (area/distance labels) were not appearing. I found that the root cause was an infinite loop where the main map component (`Mapbox.tsx`) would destroy and re-create itself whenever drawing data was updated.

The loop occurred because:
1. The `updateMeasurementLabels` function (which performs calculations) updates `mapData.drawnFeatures` in the React context.
2. `mapData.drawnFeatures` was previously included in the dependency array of the main `useEffect` hook responsible for map creation and cleanup in `Mapbox.tsx`.
3. This caused the `useEffect` to re-run its cleanup (destroying the map) and setup (creating a new map) phases every time a drawing was made or updated, interrupting the calculation and label rendering process.

The fix involves:
- Removing `mapData.drawnFeatures` from the dependency array of the main map initialization `useEffect` in `Mapbox.tsx`.

This change ensures that updating drawing features in the context does not trigger a full map rebuild. The map instance now remains stable during drawing operations, allowing `updateMeasurementLabels` to execute correctly and display the calculations. This also enables proper testing of drawing state preservation when you switch views.

The project builds successfully with these changes. The ESLint `exhaustive-deps` warning for intentionally omitted dependencies (`viewport`, `mapData.drawnFeatures`) in this specific `useEffect` is acknowledged as a necessary trade-off for these fixes.
